corehq/apps/sso/tests/test_async_handlers.py

```diff
@@ -0,0 +1,315 @@
+            {
+                'linkedObjects': [],
+            }
+        )
+
+
+class TestSSOExemptUsersAdminAsyncHandler(BaseAsyncHandlerTest):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.email_domain = AuthenticatedEmailDomain.objects.create(
+            identity_provider=cls.idp_one,
+            email_domain='vaultwax.com',
+        )
+        cls.email_domain_two = AuthenticatedEmailDomain.objects.create(
+            identity_provider=cls.idp_two,
+            email_domain='vwx.link',
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        AuthenticatedEmailDomain.objects.all().delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        super().setUp()
+        self.request = RequestFactory().get('/sso/test')
+        self.request.method = 'POST'
+        self.idp_one.is_editable = False
+        self.idp_one.save()
+
+    def tearDown(self):
+        UserExemptFromSingleSignOn.objects.all().delete()
+        super().tearDown()
+
+    def test_get_linked_objects(self):
+        UserExemptFromSingleSignOn.objects.create(
+            email_domain=self.email_domain,
+            username='b@vaultwax.com'
+        )
+        UserExemptFromSingleSignOn.objects.create(
+            email_domain=self.email_domain,
+            username='c@vaultwax.com'
+        )
+        self.request.POST = {
+            'requestContext[idpSlug]': self.idp_one.slug,
+        }
+        handler = SSOExemptUsersAdminAsyncHandler(self.request)
+        self.assertEqual(
+            handler.get_linked_objects(),
+            [
+                'b@vaultwax.com',
+                'c@vaultwax.com',
+            ]
+        )
+
+    def test_missing_email_domain_in_username_raises_error(self):
+        self.request.POST = {
+            'requestContext[idpSlug]': self.idp_one.slug,
+            'objectName': 'bademail',
+        }
+        handler = SSOExemptUsersAdminAsyncHandler(self.request)
+        with self.assertRaises(AsyncHandlerError):
+            handler.email_domain
+
+    def test_add_object_raises_errors_if_username_exists(self):
+        UserExemptFromSingleSignOn.objects.create(
+            username='b@vaultwax.com',
+            email_domain=self.email_domain
+        )
+        self.request.POST = {
+            'requestContext[idpSlug]': self.idp_one.slug,
+            'objectName': 'b@vaultwax.com',
+        }
+        handler = SSOExemptUsersAdminAsyncHandler(self.request)
+        with self.assertRaises(AsyncHandlerError):
+            handler.add_object()
+
+    def test_add_object_raises_errors_if_email_domain_is_linked_to_different_idp(self):
+        self.request.POST = {
+            'requestContext[idpSlug]': self.idp_one.slug,
+            'objectName': 'b@vwx.link',
+        }
+        handler = SSOExemptUsersAdminAsyncHandler(self.request)
+        with self.assertRaises(AsyncHandlerError):
+            handler.add_object()
```

Comment:
I want the idp to be guaranteed fresh each time because I do make modifications to it (`is_active=True/False` etc) in certain tests.

Reply:
I think I misread this. for this test the idp is created in `setUpClass`, and I read `setUp`. Generally, if I don't have to duplicate database calls I try to avoid it. I think @millerdev has said in the past `setUpClass` is always preferred if it's not necessary to have fresh data for each test.

Reply:
Curious to hear @millerdev 's thoughts, but I'd consider that premature optimization.  We're essentially talking about trading readability for speed, right? I think that a test that contains all of its phases ([the AAA pattern](https://freecontent.manning.com/making-better-unit-tests-part-1-the-aaa-pattern/)) in its body is so much nicer to deal with than one where I need to hunt through setup or subclasses. While that link does mention extracting common elements into a constructor/setup method, it also mentions:

"The other drawback to extracting the arrangement code into the constructor is diminished test readability. You no longer see the full picture by looking at the test. You have to examine different places in the class to understand what the test method does.

Even if there’s not much arrangement logic, say, only instantiation of the fixtures, you’re still better off moving it directly to the test method. Otherwise, you’ll be wondering if it’s instantiation or there’s something else being configured there too. A self-contained test doesn’t leave you with such uncertainties."

Reply:
yes, I do get where you're coming from. From that point of view, with tests it's better to be verbose rather than optimized with the goal of improved readability. However, I believe we take the over-optimized route because we also have a situation where our tests take forever to run and every millisecond adds up (limited Travis jobs). As a result, the tendency is to overly optimize test code and forgo some instant readability. Perhaps what's better is more explicit documentation/comments? also curious to hear Daniel's thoughts